### PR TITLE
major refactor

### LINKED ISSUE #60: flags.Lookup("nodename") does not work
We are being smart on how --nodename is used, but looking up the flag. But that code does not seem to be executed because the `flags.Lookup()` func is looking in the wrong flag set. We need to look this up in the original VK flagset, but that's hidden from us in cli.Command. That type doesn't give us access to `corba.Command` which does have method to access the flagset.

This also means we can't remove flags from the original set even though they are useless in systemk - they still show up in the --help :-1:

### PR DESCRIPTION
**Pull request by pires**
Fixes #60
Closes #58
Refs #48

Here's a non-exhaustive list of changes:
- drop node-cli depthat was dragging us from adopting newer VK
- disabled metrics until VK supports k8s.io/kubelet v0.20+
- control flags
- improved IP addresses handling
- set-up VK logging
- use VK klogv2 structured logging facility
- when OS package management is not possible, default to no-op
- enable event recording for Pod controller
- GOARCH is now used to populate Node's architecture label
- improved Pod-related resource handling

### CODE REVIEW

**Comment by miekg** [comment]
> it seems NodeAddresses only support one InternalIP and one ExternalIP, 

I believe there is also `*IPs` that contains a list, which isn't a real list, but a tuple that also holds the IPv6 address

File: cmd/http.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+		}
+	}
+	defer func() {
+		if retErr != nil {
+			cancel()
+		}
+	}()
+
+	if cfg.CertPath == "" || cfg.KeyPath == "" {
+		log.G(ctx).
```
Comment (miekg) [inline cmd/http.go:76]: please fix the logging here is well, `G(ctx)` is not nice to type, also not a fan a "structured" logging

File: cmd/http.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+			Handler:   mux,
+			TLSConfig: tlsCfg,
+		}
+		go serveHTTP(ctx, s, l, "pods")
+		closers = append(closers, s)
+	}
+
+	if cfg.MetricsAddr == "" {
+		log.G(ctx).Info("pod metrics server not setup due to empty metrics address")
+	} else {
```
Comment (miekg) [inline cmd/http.go:113]: no need for else you can just return here, and the dedent the rest of the funciton

File: cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
+// Copyright © 2017 The virtual-kubelet authors
```
Comment (miekg) [inline cmd/root.go:1]: do we really need all this? We want a binary with some flags that implements the Provider interface? That would just be some `cobra` stuff

File: internal/kubernetes/watch.go [hunk 4]
```diff
@@ -28,29 +62,49 @@ type watcher struct {
...
 		secrets:         make(map[types.NamespacedName][]*corev1.Pod),
 		secretKeysByPod: make(map[types.NamespacedName][]types.NamespacedName),
+		cmLister:        informerFactory.Core().V1().ConfigMaps().Lister(),
+		secretLister:    informerFactory.Core().V1().Secrets().Lister(),
 	}
 }
 
-func (w *watcher) handlerFuncs(ctx context.Context, p updater) cache.ResourceEventHandlerFuncs {
+func (w *watcher) ConfigMapLister() listersv1.ConfigMapLister {
+	return w.cmLister
```
Comment (miekg) [inline internal/kubernetes/watch.go:88]: Just make these fields public, no need for 1-line accesor functions

File: internal/ospkg/archlinux.go [hunk 5]
```diff
@@ -1,4 +1,18 @@
-package packages
+// Copyright © 2021 The systemk authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (miekg) [inline internal/ospkg/archlinux.go:3]: not every file needs a copyright. In fact this only needs to be done one in the License file

File: internal/ospkg/archlinux.go [hunk 6]
```diff
@@ -9,25 +23,25 @@ import (
 	"path"
 	"strings"
 
-	"github.com/virtual-kubelet/systemk/pkg/unit"
+	"github.com/virtual-kubelet/systemk/internal/unit"
```
Comment (miekg) [inline internal/ospkg/archlinux.go:26]: yes `internal` exists as a thing, but now I need to `cd` once more. Can we remove `internal`?

File: internal/provider/node.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+		"memory":  resource.MustParse(system.Memory()),
+		"pods":    resource.MustParse(system.Pid()),
+		"storage": resource.MustParse("40G"), // needs the size of the /var FS
+	}
+}
+
+func nodeConditions() []corev1.NodeCondition {
+	return []corev1.NodeCondition{
+		// Don't set the Ready condition to True just yet.
+		// That shall happen when all of systemk has been initialized.
```
Comment (miekg) [inline internal/provider/node.go:153]: can we point to a go file/function where that happens?

File: internal/provider/provider.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provider
+
+import (
+	"context"
+	"github.com/virtual-kubelet/systemk/internal/kubernetes"
```
Comment (miekg) [inline internal/provider/provider.go:19]: import sorting

File: internal/provider/provider.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	"os"
+
+	"github.com/virtual-kubelet/systemk/internal/ospkg"
+	"github.com/virtual-kubelet/systemk/internal/system"
+	"github.com/virtual-kubelet/systemk/internal/unit"
+	"github.com/virtual-kubelet/virtual-kubelet/node"
+	"github.com/virtual-kubelet/virtual-kubelet/node/api"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	stats "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
```
Comment (miekg) [inline internal/provider/provider.go:30]: if possible lets not pull this in, a rather *not* have this functionality than pulling in k8s.io/kubernetes

File: internal/unit/file.go [hunk 10]
```diff
@@ -23,8 +25,8 @@ import (
 	"github.com/coreos/go-systemd/v22/unit"
 )
 
-// New returns a new unit file parsed from raw.
-func New(raw string) (*File, error) {
+// NewUnitFile returns a new unit file parsed from raw.
+func NewUnitFile(raw string) (*File, error) {
```
Comment (miekg) [inline internal/unit/file.go:29]: `New` is fine here, `unit.New`, `unit.NewUnitFile` stotters

File: internal/util/util.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
+// Copyright © 2021 The systemk authors
```
Comment (miekg) [inline internal/util/util.go:1]: please rename `util` to something more meaningful

File: main.go [hunk 12]
```diff
@@ -16,102 +16,51 @@ package main
...
+	"github.com/virtual-kubelet/systemk/internal/provider"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	"github.com/virtual-kubelet/virtual-kubelet/log/klogv2"
 )
 
 var (
 	buildVersion = "N/A"
 	buildTime    = "N/A"
-	k8sVersion   = "v1.18.4" // Inject this build time by parsing mod.go
+	k8sVersion   = "v1.18.15" // This should follow the version of k8s.io/kubernetes we are importing
```
Comment (miekg) [inline main.go:35]: I think this can be removed as we don't wont to be importaint k8s.io/kubernetes

File: main.go [hunk 13]
```diff
@@ -16,102 +16,51 @@ package main
...
-	ctx := cli.ContextWithCancelOnSignal(context.Background())
+	// Setup a local logger.
+	logger := klogv2.New(nil)
 
-	o, err := opts.FromEnv()
-	if err != nil {
-		klog.Fatal(err)
+	var opts provider.Opts
+	if err := provider.SetDefaultOpts(&opts); err != nil {
+		logger.Fatalf("failed to process options: %s", err)
```
Comment (miekg) [inline main.go:55]: I think logger, can just be `log` or even `klog`

**Comment by miekg** [comment]
slew of comments, but very much in favor of this change!

File: internal/kubernetes/watch.go (hunk 4 above)
Comment (pires) [reply internal/kubernetes/watch.go:88]: That doesn't work when ctors return an interface impl. If I change the ctor, then I cannot use the interface to pass this to provider. I wanted to separate what is pure Kubernetes stuff like this from what is provider.

File: cmd/http.go (hunk 1 above)
Comment (pires) [reply cmd/http.go:76]: Logging would be done as part of #58 but if you prefer, I'll do it here.

**Comment by miekg** [comment]
ok! fair enough

File: cmd/http.go (hunk 2 above)
Comment (pires) [reply cmd/http.go:113]: This is maybe the file I didn't spend time on. Will do!

File: cmd/root.go (hunk 3 above)
Comment (pires) [reply cmd/root.go:1]: Implementing the Provider interface is not enough, and that's why node-cli came to be. So yes, this code is very much needed and until node-cli is reworked as part of VK into pieces you can pick, I can't see other option other than having this here.

File: internal/ospkg/archlinux.go (hunk 5 above)
Comment (pires) [reply internal/ospkg/archlinux.go:3]: Sorry, this is a thing I have for license headers, and usually, in other projects, I do have automation to put it into place. If this is something you're strongly opinionated about, I can remove all headers on files fully created by us.

File: internal/provider/node.go (hunk 7 above)
Comment (pires) [reply internal/provider/node.go:153]: It's in root.go, as the last step after Node and Pod controllers are running. https://github.com/virtual-kubelet/systemk/pull/65/files#diff-ab967ab1a2f3a1b769106eeb7bfe892ef0e81d1d27811fa15be08e6749feee1fR209

File: internal/provider/provider.go (hunk 8 above)
Comment (pires) [reply internal/provider/provider.go:19]: Good catch!

File: internal/provider/provider.go (hunk 9 above)
Comment (pires) [reply internal/provider/provider.go:30]: OK then we'll have to pull this one out until VK pulls in `k8s.io/kubelet v0.20.2` of newer.

File: internal/unit/file.go (hunk 10 above)
Comment (pires) [reply internal/unit/file.go:29]: True but maybe `NewFile` then? First, in this package you also find `NewManager` for a new `UnitManager`, and second this func doesn't return a unit but a unit file.

File: internal/util/util.go (hunk 11 above)
Comment (pires) [reply internal/util/util.go:1]: Suggestions? These are utils used in code scattered in other packages, including tests.

File: main.go (hunk 12 above)
Comment (pires) [reply main.go:35]: Well, users will need to know if systemk is compatible w/ their Kubernetes cluster. For instance, this won't work w/ a Kubernetes <1.20 as soon as you import `k8s.io/kubelet v0.20.2` for providing Pod metrics.

I think we should abide to https://kubernetes.io/docs/setup/release/version-skew-policy/#kubelet.

File: main.go (hunk 13 above)
Comment (pires) [reply main.go:55]: I usually don't like having var names that collide with package names but I'm not strong about the name of the logger. Anyways, as mentioned before, I didn't rework logging as part of this PR but #58 instead. I'm fine doing it here.

File: cmd/root.go (hunk 3 above)
Comment (pires) [reply cmd/root.go:1]: Miek's gave a nod to my last comment https://github.com/virtual-kubelet/systemk/pull/65#issuecomment-763473080

File: cmd/http.go (hunk 1 above)
Comment (pires) [reply cmd/http.go:76]: As discussed over slack, I'll do this here too.

File: cmd/http.go (hunk 2 above)
Comment (pires) [reply cmd/http.go:113]: Actually, this needs to go away so we don't import k8s.io/kubernetes.

File: internal/kubernetes/watch.go (hunk 4 above)
Comment (pires) [reply internal/kubernetes/watch.go:88]: Miek's gave a nod to my last comment https://github.com/virtual-kubelet/systemk/pull/65#issuecomment-763473080

File: internal/provider/provider.go (hunk 8 above)
Comment (pires) [reply internal/provider/provider.go:19]: Done - will push soon.

File: main.go (hunk 12 above)
Comment (pires) [reply main.go:35]: Ah if by "importaint" you didn't mean important but importing, the comment is misleading but my comment above stands.

File: internal/ospkg/archlinux.go (hunk 5 above)
Comment (miekg) [reply internal/ospkg/archlinux.go:3]: yes it's not needed, if you can please remove

File: internal/provider/node.go (hunk 7 above)
Comment (miekg) [reply internal/provider/node.go:153]: please add this to the comment

File: internal/provider/provider.go (hunk 9 above)
Comment (miekg) [reply internal/provider/provider.go:30]: yes please

File: internal/unit/file.go (hunk 10 above)
Comment (miekg) [reply internal/unit/file.go:29]: `NewFile` sgtm

File: main.go (hunk 12 above)
Comment (miekg) [reply main.go:35]: We should be careful by providing those features that are so tied to a k8s version esp if they are so new. Why not focus first on the plain podspec integration and worry about this later when we actually want and need this?

File: main.go (hunk 13 above)
Comment (miekg) [reply main.go:55]: I like short names and `klog`, or `log` is short and to the point

File: internal/ospkg/archlinux.go (hunk 5 above)
Comment (pires) [reply internal/ospkg/archlinux.go:3]: Will do.

File: cmd/http.go (hunk 2 above)
Comment (pires) [reply cmd/http.go]: Done but let comments to make it easier to revert when VK addresses this import.

File: internal/ospkg/archlinux.go (hunk 5 above)
Comment (pires) [reply internal/ospkg/archlinux.go]: Done!

File: internal/provider/provider.go (hunk 9 above)
Comment (pires) [reply internal/provider/provider.go]: Done!

File: internal/unit/file.go (hunk 10 above)
Comment (pires) [reply internal/unit/file.go]: Done!

File: main.go (hunk 12 above)
Comment (pires) [reply main.go:34]: This is about the kubelet API not the core API. Pod Stats is a kubelet API.

File: main.go (hunk 13 above)
Comment (pires) [reply main.go]: Done!

**Comment by pires** [comment]
Currently running logs, debugging disabled
```
k# ./systemk --kubeconfig systemk.kubeconfig --nodename systemk --internal-ip 192.168.1.123 --external-ip 9.0.0.7
I0120 14:00:19.688249   40734 provider.go:90] installing policyrcd-script-zg2, to prevent installed daemons from starting
I0120 14:00:19.688562   40734 debian.go:29] checking if "policyrcd-script-zg2" is installed [os=debian]
E0120 14:00:19.912946   40734 http.go:78] TLS certificates are required to serve the kubelet API [cert= key=]
I0120 14:00:19.915950   40734 podcontroller.go:287] Pod cache in-sync [source=virtual-kubelet]
I0120 14:00:19.917291   40734 root.go:205] systemk initialized [node=systemk]
I0120 14:00:19.921882   40734 podcontroller.go:368] starting workers [source=virtual-kubelet]
I0120 14:00:19.932786   40734 podcontroller.go:380] started workers [source=virtual-kubelet]
I0120 14:00:27.160473   40734 pod.go:72] CreatePod called [podName=hello podNamespace=default]
I0120 14:00:27.161024   40734 pod.go:88] processing container 0 (init=false) [dir=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0 file=namespace podName=hello podNamespace=default]
I0120 14:00:27.161294   40734 debian.go:29] checking if "bash" is installed [os=debian]
W0120 14:00:27.175133   40734 unit.go:287] failed to find unit file, synthesizing one [dir=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0 file=namespace podName=hello podNamespace=default]
I0120 14:00:27.176138   40734 pod.go:239] loading unit bash, bash as systemk.default.hello.bash.service
[Unit]
Description=systemk
Documentation=man:systemk(8)

[Install]
WantedBy=multi-user.target

[Service]
ProtectSystem=true
ProtectHome=tmpfs
PrivateMounts=true
ReadOnlyPaths=/
StandardOutput=journal
StandardError=journal
RemainAfterExit=true
ExecStart=/bin/bash -c "while true; do echo hello; sleep 1; done"
TemporaryFileSystem=/var /run
BindReadOnlyPaths=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0:/var/run/secrets/kubernetes.io/serviceaccount
Environment=HOSTNAME=systemk
Environment=KUBERNETES_SERVICE_PORT=6444
Environment=KUBERNETES_SERVICE_HOST=127.0.0.1
Environment=SYSTEMK_NODE_INTERNAL_IP=192.168.1.123
Environment=SYSTEMK_NODE_EXTERNAL_IP=9.0.0.7

[X-Kubernetes]
Namespace=default
ClusterName=
Id=3de49133-8c57-4112-a861-f717bb2ec7e5
 [dir=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0 file=namespace podName=hello podNamespace=default]
I0120 14:00:27.181636   40734 manager.go:249] writing systemd unit systemk.default.hello.bash.service:
 (759b)
I0120 14:00:27.183013   40734 manager.go:262] enabling systemd unit systemk.default.hello.bash.service
I0120 14:00:27.184452   40734 pod.go:249] starting unit "systemk.default.hello.bash.service" [dir=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0 file=namespace podName=hello podNamespace=default]
I0120 14:00:27.190893   40734 manager.go:121] triggered systemd unit systemk.default.hello.bash.service start: job=11569
I0120 14:00:27.191312   40734 pod.go:103] Created pod in provider [source=virtual-kubelet]
I0120 14:00:35.311459   40734 pod.go:321] DeletePod called [dir=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0 file=namespace podName=hello podNamespace=default]
I0120 14:00:35.317615   40734 manager.go:271] removing systemd unit systemk.default.hello.bash.service
W0120 14:00:35.320570   40734 pod.go:334] failed to unload unit "systemk.default.hello.bash.service": <nil>, Unit systemk.default.hello.bash.service not loaded. [dir=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0 file=namespace podName=hello podNamespace=default]
I0120 14:00:35.321207   40734 pod.go:336] deleted unit "systemk.default.hello.bash.service" successfully [dir=/var/run/3de49133-8c57-4112-a861-f717bb2ec7e5/secrets/#0 file=namespace podName=hello podNamespace=default]
^CI0120 14:01:13.505162   40734 main.go:67] system exited gracefully
```

File: internal/provider/node.go (hunk 7 above)
Comment (pires) [reply internal/provider/node.go:145]: Done!

**Comment by pires** [comment]
Node description
```
Name:               systemk
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/os=Linux
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=systemk
                    kubernetes.io/os=Linux
                    node.kubernetes.io/instance-type=systemk
Annotations:        node.alpha.kubernetes.io/ttl: 0
                    virtual-kubelet.io/last-applied-node-status:
                      {"capacity":{"cpu":"1","memory":"1004912k","pods":"1048576","storage":"40G"},"allocatable":{"cpu":"1","memory":"1004912k","pods":"1048576"...
                    virtual-kubelet.io/last-applied-object-meta:
                      {"name":"systemk","uid":"d0091f8b-f460-43ed-a469-07cc36880a0b","creationTimestamp":null,"labels":{"kubernetes.io/arch":"amd64","kubernetes...
CreationTimestamp:  Wed, 20 Jan 2021 14:32:28 +0000
Taints:             virtual-kubelet.io/provider=systemk:NoSchedule
Unschedulable:      false
Lease:
  HolderIdentity:  systemk
  AcquireTime:     <unset>
  RenewTime:       Wed, 20 Jan 2021 14:32:28 +0000
Conditions:
  Type                 Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----                 ------  -----------------                 ------------------                ------                       -------
  Ready                True    Wed, 20 Jan 2021 14:32:28 +0000   Wed, 20 Jan 2021 14:32:28 +0000   KubeletReady                 systemk is ready
  OutOfDisk            False   Wed, 20 Jan 2021 14:32:28 +0000   Wed, 20 Jan 2021 14:32:28 +0000   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure       False   Wed, 20 Jan 2021 14:32:28 +0000   Wed, 20 Jan 2021 14:32:28 +0000   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure         False   Wed, 20 Jan 2021 14:32:28 +0000   Wed, 20 Jan 2021 14:32:28 +0000   KubeletHasNoDiskPressure     kubelet has no disk pressure
  NetworkUnavailable   False   Wed, 20 Jan 2021 14:32:28 +0000   Wed, 20 Jan 2021 14:32:28 +0000   RouteCreated                 RouteController created a route
  PIDPressure          False   Wed, 20 Jan 2021 14:32:28 +0000   Wed, 20 Jan 2021 14:32:28 +0000   KubeletHasSufficientPIDs     kubelet has sufficient PIDs available
Addresses:
  InternalIP:  192.168.64.15
  ExternalIP:  192.168.64.15
Capacity:
  cpu:      1
  memory:   1004912k
  pods:     1048576
  storage:  40G
Allocatable:
  cpu:      1
  memory:   1004912k
  pods:     1048576
  storage:  40G
System Info:
  Machine ID:
  System UUID:
  Boot ID:
  Kernel Version:             5.4.0-60-generic
  OS Image:                   Ubuntu 20.04.1 LTS
  Operating System:           Linux
  Architecture:               amd64
  Container Runtime Version:  systemd 245 (245.4-4ubuntu3.3)
  Kubelet Version:            v1.18.15
  Kube-Proxy Version:
PodCIDR:                      10.244.6.0/24
PodCIDRs:                     10.244.6.0/24
Non-terminated Pods:          (0 in total)
  Namespace                   Name    CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                   ----    ------------  ----------  ---------------  -------------  ---
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests  Limits
  --------           --------  ------
  cpu                0 (0%)    0 (0%)
  memory             0 (0%)    0 (0%)
  ephemeral-storage  0 (0%)    0 (0%)
  storage            0         0
Events:              <none>
```

File: internal/util/util.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+// GetEnvOrDefault returns environment variable value or if unset, a default value.
+func GetEnvOrDefault(key, defaultValue string) string {
```
Comment (miekg) [inline internal/util/util.go:13]: this is only used in internal/provider, just move it there, this means this package can then be renamed to podutil and the PodFromFile function can be renamed to `FromFile`

Comment (pires) [reply internal/util/util.go:13]: Gotcha!

**Comment by miekg** [comment]
if you have it working locally we can prolly merge pretty soon and fix anything that needs fixing later on.

**Comment by pires** [comment]
I should have this ready to go tomorrow sometime in our morning time.

File: internal/util/util.go (hunk 14 above)
Comment (pires) [reply internal/util/util.go:13]: Done!

**Comment by pires** [comment]
This is good to go!

**Comment by pires** [comment]
This is good to go, @miekg!

**Pull request merged**